### PR TITLE
Add ReAct baseline

### PR DESCRIPTION
**Pull request by paulnovello**
**ReAct** (Sun et al, 2021): _"Out-of-distribution Detection With Rectified Activations"_ ([link](https://arxiv.org/pdf/2111.12797.pdf))
---

This baseline is particular because it applies a threshold to the penultimate layer's activations so it is compatible with any other OOD scores.

It is conceptually very simple but involves thinking carefully about its implementation. The challenges are:
* How to seamlessly integrate it into all other metrics without modifying them too much? -> Maybe put it into `base.py`
* It involves the computation of a percentile, so a forward. How to avoid duplicate forward for baselines that then use `_fit_dataset`? -> Maybe save the activation values when using the feature extractor and allow starting from these (possibly thresholded) features.
* How to efficiently modify the computation graph to threshold the activation values?

### CODE REVIEW

**Comment by y-prudent** [comment]
## WIP

- [x] How to efficiently modify the computation graph to threshold the activation values?
    **Solution proposed**
    - In Torch, this is done using `forward_hooks` to clip the output of the penultimate layer
    - In Tensorflow the extractor is split into two parts (input -> penultimate layer / after penultimate layer -> last layer) using the `tf.keras.model.Model` class.
- [ ] It involves the computation of a percentile, so a forward. How to avoid duplicate forward for baselines that then use `_fit_dataset`?
    **TODO**

**Comment by y-prudent** [comment]
Ready to be reviewed ! @cofri @paulnovello 
I am sorry in advance the commits are a bit messy and it was difficult to rearrange them using an interactive rebase because some of them are entangled (many refactorings). Please refer to the previous message to see which are the main features that need to be checked.

File: oodeel/extractor/keras_feature_extractor.py [hunk 1]
```diff
@@ -83,10 +95,18 @@ def find_layer(self, layer_id: Union[str, int]) -> tf.keras.layers.Layer:
...
         """
         if isinstance(layer_id, str):
-            return self.model.get_layer(layer_id)
+            layers_names = [layer.name for layer in model.layers]
+            layer_id = layers_names.index(layer_id)
         if isinstance(layer_id, int):
-            return self.model.get_layer(index=layer_id)
-        raise ValueError(f"Could not find any layer {layer_id}.")
+            layer_id += index_offset
+            layer = model.get_layer(index=layer_id)
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply oodeel/extractor/keras_feature_extractor.py:102]: This line is correctly indented: If `layer_id` is a string, then it is converted to an integer through the first `if` condition, so the index offset can then be applied through the second condition.

[the remark this answers is not in the document]
Comment (cofri) [reply oodeel/extractor/keras_feature_extractor.py:102]: My bad! I was mistaken with the "if" and not "elif".

File: oodeel/methods/energy.py [hunk 2]
```diff
@@ -61,10 +77,18 @@ def _score_tensor(self, inputs: TensorType) -> np.ndarray:
...
         return scores
+
+    def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
+        """
+        Fits the OOD detector to fit_dataset.
+
+        Args:
+            fit_dataset: dataset to fit the OOD detector on
+        """
+        pass
```
Comment (paulnovello) [inline oodeel/methods/energy.py:94]: Maybe add a comment explaining why the method is empty?

File: oodeel/methods/entropy.py [hunk 3]
```diff
@@ -63,3 +78,12 @@ def _score_tensor(self, inputs: TensorType) -> np.ndarray:
...
         return -scores
+
+    def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
+        """
+        Fits the OOD detector to fit_dataset.
+
+        Args:
+            fit_dataset: dataset to fit the OOD detector on
+        """
+        pass
```
Comment (paulnovello) [inline oodeel/methods/entropy.py:89]: Maybe add a comment explaining why the method is empty?

File: oodeel/methods/mls.py [hunk 4]
```diff
@@ -65,3 +79,12 @@ def _score_tensor(self, inputs: TensorType) -> np.ndarray:
...
         return scores
+
+    def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
+        """
+        Fits the OOD detector to fit_dataset.
+
+        Args:
+            fit_dataset: dataset to fit the OOD detector on
+        """
+        pass
```
Comment (paulnovello) [inline oodeel/methods/mls.py:90]: Maybe add a comment explaining why the method is empty?

File: oodeel/methods/odin.py [hunk 5]
```diff
@@ -95,3 +107,12 @@ def _temperature_loss(self, inputs: TensorType, labels: TensorType) -> TensorTyp
...
         return loss
+
+    def _fit_to_dataset(self, fit_dataset: DatasetType) -> None:
+        """
+        Fits the OOD detector to fit_dataset.
+
+        Args:
+            fit_dataset: dataset to fit the OOD detector on
+        """
+        pass
```
Comment (paulnovello) [inline oodeel/methods/odin.py:118]: Maybe add a comment explaining why the method is empty?

**Review by paulnovello — COMMENTED** [review]
Maybe we should add a test on react that checks that the intermediate values are actually thresholded to the desired quantile ?

File: oodeel/extractor/torch_feature_extractor.py [hunk 6]
```diff
@@ -93,29 +98,56 @@ def hook(_, __, output):
 
         return hook
 
-    def find_layer(self, layer_id: Union[str, int]) -> nn.Module:
+    @staticmethod
+    def find_layer(
+        model: nn.Module,
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply oodeel/extractor/torch_feature_extractor.py:103]: `find_layer` is defined as a static method, with the model passed as an argument in order to call it in `OODBaseDetector.compute_react_threshold` to get the `penultimate_layer_id` before instanciating the extractor for the penultimate features.

**Comment by y-prudent** [comment]
> Maybe we should add a test on react that checks that the intermediate values are actually thresholded to the desired quantile? Very good otherwise, good job!

Thanks! And good idea for the treshold test, I just added it. Despite the appearances, it was a bit tricky without modifying source code since torch and keras extractors are asymetrical (keras splits the model and apply the clipping operation between penultimate and ultimate layer whereas torch adds a hook on the penultimate layer). I found a solution for both frameworks but the react test should always be done at the end of the `eval_detector_on_blobs` since it may modify the detector's behaviour.

**Review by cofri — APPROVED** [review]

**Pull request merged**